app/src/main/java/ch/hikemate/app/model/elevation/ElevationServiceRepository.kt

```diff
@@ -1,113 +1,270 @@
+) : ElevationRepository {
+
+  // The number of failed requests made to the Elevation API since the last successful request
+  private var failedRequests = 0
 
   // Cache for the elevation data, indexed by hike ID
-  private val cache = mutableMapOf<String, List<Double>>()
+  private val cache = hashMapOf<LatLong, ElevationCacheEntry>()
+
+  // List of requests that are waiting for the cache to be updated
+  private val requests = mutableListOf<ElevationRequest>()
+
+  // A mutex to ensure that the cache and requests are accessed atomically
+  private val mutex: Mutex = Mutex()
+
+  // The job performing the cache update
+  private var cacheUpdateJob: Job? = null
 
   /**
    * Get the size of the cache
    *
    * @return The size of the cache
    */
-  fun getCacheSize(): Int {
+  private fun getCacheSize(): Int {
     return cache.size
   }
 
   override fun getElevation(
       coordinates: List<LatLong>,
-      hikeID: String,
       onSuccess: (List<Double>) -> Unit,
       onFailure: (Exception) -> Unit
   ) {
     if (coordinates.isEmpty()) {
       onSuccess(emptyList())
       return
     }
-    // If the cache already has the elevation data for this hike, call onSuccess
-    if (cache.containsKey(hikeID)) {
-      cache[hikeID]?.let { onSuccess(it) }
+
+    if (coordinates.minus(cache.keys).isEmpty()) {
+      // All coordinates are in the cache
+      onSuccess(
+          coordinates.map {
+            val value = cache[it]
+            if (value == null) {
+              Log.e("ElevationServiceRepository", "Coordinate is missing in cache")
+              return@map 0.0
+            } else {
+              value.elevation
+            }
+          })
       return
     }
 
-    // Check if a request is already in progress
-    if (cache[hikeID] == null) {
-      cache[hikeID] = emptyList()
-    } else {
-      return
+    CoroutineScope(dispatcher).launch {
+      mutex.withLock {
+        requests.add(ElevationRequest(coordinates, onSuccess, onFailure))
+
+        if (cacheUpdateJob == null || cacheUpdateJob?.isCompleted == true) {
+          launchUpdateOfCache()
+        }
+      }
     }
+  }
 
-    // Create the JSON body for the request
-    val jsonBody = buildJsonObject {
-      // The major object
-      put(
-          LOCATION_KEY,
-          buildJsonArray {
-            coordinates.forEach { coordinate ->
-              add(
-                  // Object with the location
-                  buildJsonObject {
-                    put(LATITUDE_KEY, coordinate.lat)
-                    put(LONGITUDE_KEY, coordinate.lon)
-                  })
+  /**
+   * Launch the update of the cache
+   *
+   * The update will be delayed by 1 second to allow for multiple requests to be batched together
+   *
+   * @param onSuccess The callback to be called when the request is successful
+   * @param onFailure The callback to be called when the request fails
+   */
+  private fun launchUpdateOfCache(
```

Comment:
Actually, it acquires the lock! It does not before the emptiness check, because this method should not be called on an empty request list, but this is more an assertion than an iteration on the data. If new requests are added after (and it should be in a suspending way), then the code will recall this function! 

The only issue I see is if `requests` were non-empty, then it passes this check, and requests can be cleared after. However, this is not an issue since the coordinates set will be empty and thus the for-loop will exit early.

![image](https://github.com/user-attachments/assets/31494114-8efe-4b20-82b0-4537473d00a8)

Reply:
Yup, I re-read the function and it looks good to me, I think I got lost the first time with all the locks haha